### PR TITLE
Check available index versions for Delta Lake time travel query

### LINKED ISSUE #387: Check removed index log versions of time travel query index candidates
Currently, time travel query optimization assumes all index log versions exist. 
However, as most of index applications are based on the latest index log version, some of outdated version might be deleted by a user. 
Therefore, we need to check the removed log versions and filter them out in the candidate log version list in `closestIndexVersion`.

ref:
https://github.com/microsoft/hyperspace/pull/272#r598211313_

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?

 - **Tracking Issue**: n/a
 - **Parent Issue**:  n/a
 - **Dependencies**: n/a

Follow up PR -  https://github.com/microsoft/hyperspace/pull/272#discussion_r598329940

Fixes #387

### What changes were proposed in this pull request?

Check the existence of index data directories for each index log version before using them in Delta Lake time travel query. 
To do so, added below APIs:
```scala
// in IndexLogManager
  /** Returns index log versions whose state is in the given states */
  def getIndexVersions(states: Seq[String]): Seq[Int]

// in IndexManager
/**
   * Get index log version ids of the given index that match any of the given states.
   *
   * @param indexName Name of the index.
   * @param states List of index states of interest.
   * @return Index log versions.
   */
  def getIndexVersions(indexName: String, states: Seq[String]): Seq[Int]
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test

### CODE REVIEW

File: src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala [hunk 1]
```diff
@@ -162,12 +180,6 @@ class IndexCollectionManager(
     }
   }
 
-  override def getIndex(indexName: String, logVersion: Int): Option[IndexLogEntry] = {
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala:165]: note: moved to above

File: src/main/scala/com/microsoft/hyperspace/Hyperspace.scala [hunk 2]
```diff
@@ -163,6 +163,17 @@ class Hyperspace(spark: SparkSession) {
...
   }
+
+  /**
+   * Get index metadata and detailed index statistics for a given index and log version.
+   *
+   * @param indexName Name of the index to get stats for.
+   * @param logVersion Index log version to retrieve.
+   * @return Index metadata and statistics as a [[DataFrame]].
+   */
+  def index(indexName: String, logVersion: Int): DataFrame = {
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/Hyperspace.scala:174]: https://github.com/microsoft/hyperspace/pull/272#discussion_r598329940 was actually asking to add an API to `IndexManager` to return available index versions. For example, `def getAvailableIndexVersions(indexName: String): Seq[Int]` so that versions can be used for `def getIndex(indexName: String, logVersion: Int)`. Does it make sense?

We can also update the existing `def index(indexName: String): DataFrame` to return all available versions of index instead of exposing a new `def index(indexName: String, logVersion: Int)` (I don't think we need to expose this as public). Btw, you don't have to address this in this PR since I am not sure if we need it or not. We can just concentrate on updating `IndexLogManager` in this PR.

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/Hyperspace.scala:174]: >#272 (comment) was actually asking to add an API to IndexManager to return available index versions. For example, def getAvailableIndexVersions(indexName: String): Seq[Int] so that versions can be used for def getIndex(indexName: String, logVersion: Int). Does it make sense?

Yep then we could fix #387 with the API. I will update this PR.

>We can also update the existing def index(indexName: String): DataFrame to return all available versions of index instead of exposing a new def index(indexName: String, logVersion: Int) (I don't think we need to expose this as public). Btw, you don't have to address this in this PR since I am not sure if we need it or not. We can just concentrate on updating IndexLogManager in this PR.

I think it might be good to expose this API since we expose the log version in plan. A user might want to check the used log version.

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/Hyperspace.scala:174]: > I think it might be good to expose this API since we expose the log version in plan. A user might want to check the used log version.

If we want to expose the log versions, we can change `index` to return log entries of multiple versions (add an optional param to return last N entries). Since we return `DataFrame`, the user can just filter by the version if needed; I don't think it's necessary to expose this new API, which will require us to expose another API to return available versions.

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/Hyperspace.scala:174]: @imback82 BTW, currently if an index is deleted & vacuumed and then a new index with the same index name is created, log files are not removed & the log number is continued. I think this can be a problem. WDYT?

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/Hyperspace.scala:174]: Good point. It was intentional to keep the log files around (compliance reason), but I think we could provide an option to remove log files.

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/Hyperspace.scala:174]: Feel free to create an issue so we can track it. Thanks.

File: src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala [hunk 3]
```diff
@@ -206,7 +214,7 @@ class DeltaLakeRelation(spark: SparkSession, override val plan: LogicalRelation)
         if (equalOrLessThanLastIndex == versions.size - 1) {
           // The given table version is equal or larger than the latest index's.
           // Use the latest version.
-          index
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:209]: @imback82 `getLatestLog()` in getIndexes doesn't guarantee the availability & ACTIVE state. Should we also fix that?

File: src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala [hunk 4]
```diff
@@ -138,6 +138,24 @@ class IndexCollectionManager(
...
+  }
+
+  override def getAvailableIndexVersions(indexName: String): Seq[Int] = {
+    withLogManager(indexName) { logManager =>
+      logManager.getActiveIndexVersions().filter { id =>
+        val log = logManager.getLog(id).get
+        val indexDataPaths = log.asInstanceOf[IndexLogEntry].indexContentDirectoryPaths
+        val hadoopConf = spark.sessionState.newHadoopConf()
+        val fs = fileSystemFactory.create(new Path(indexDataPaths.head), hadoopConf)
+        indexDataPaths.forall(path => fs.exists(new Path(path)))
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala:154]: This is to handle the case where index is vacuumed? How about we move the check to `getActiveIndexVersions` where you stop iterating if you see vacuum operation?

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala [hunk 5]
```diff
@@ -44,6 +44,9 @@ trait IndexLogManager {
   /** Returns the latest LogEntry whose state is STABLE */
   def getLatestStableLog(): Option[LogEntry]
 
+  /** Returns Active index log versions */
+  def getActiveIndexVersions(): Seq[Int]
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:48]: Does it make sense to pass `states: Seq[String]` to make it more generic? `def getIndexVersions(states: Seq[String]): Seq[Int]`?

Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:48]: I noticed that we are using "id" to refer the "version". Should we change it to "id"? WDYT?

File: src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala (hunk 3 above)
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:209]: You mean https://github.com/microsoft/hyperspace/blob/40c52431b7220ed892f6781efef578af6b826fa3/src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala#L120-L126
?

For the state, it will be handled `.filter(index => states.isEmpty || states.contains(index.state))`? What's the current behavior (error message), if the index data doesn't exist?

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:209]: No errors, but just apply the invalid index version; which results in empty result.

File: src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala (hunk 4 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala:154]: It's for vacuumed or manually removed directories (by storage services).

File: src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala (hunk 5 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:48]: I think version is more general term to represent it, but the code base is `id`.. might be good to be consistent.
But in the plan, we print it as `LogVersion: ` ?

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:48]: It's because I didn't expect any use case for other states; but will fix it.

File: src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala (hunk 3 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:209]: > For the state, it will be handled .filter(index => states.isEmpty || states.contains(index.state))? What's the current behavior (error message), if the index data doesn't exist?

Yep, but it cannot return any previous version of each index if the state of the latest version is not in `states`.

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:209]: good catch. if that's the case, we should definitely fix it.

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:209]: But because of that, we could impl gold standard w/o actual data. 
I created the issue last year #206, but it's a backlog due to low importance.  (as Spark also doesn't check the file existence while reading it, though Spark check that when df creation..)
Maybe it would be good if we could check the data directories. 
@imback82 WDYT?

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:209]: Ah OK. I think just checking directories is not sufficient. We probably need to either check all files, or check nothing based on config?

> Yep, but it cannot return any previous version of each index if the state of the latest version is not in `states`.

Sorry, I didn't fully get this part. It would return `Nil` in this case, which is fine?

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:209]: > Sorry, I didn't fully get this part. It would return Nil in this case, which is fine?

if a user cancel the spark job in the middle of optimize (or refresh), the latest state remains "optimizing". 
However, we could use the previous log version if it's eligible.

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:209]: > Ah OK. I think just checking directories is not sufficient. We probably need to either check all files, or check nothing based on config?

But what about this PR? checking all files for every version should burden.

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:209]: Checking just data directory doesn't seem any better than not checking at all. I am fine with not checking at all in this PR. And we can do a follow up to introduce a config for "nocheck", "directory_check", and "file_check" in one shot.

I thought this PR can be just exposing versions, or am I missing something?

File: src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala [hunk 6]
```diff
@@ -138,6 +138,18 @@ class IndexCollectionManager(
...
 
+  override def getIndex(indexName: String, logVersion: Int): Option[IndexLogEntry] = {
+    withLogManager(indexName) { logManager =>
+      logManager.getLog(logVersion).map(_.asInstanceOf[IndexLogEntry])
+    }
+  }
+
+  override def getAvailableIndexVersions(indexName: String): Seq[Int] = {
+    withLogManager(indexName) { logManager =>
+      logManager.getIndexVersions(Seq(Constants.States.ACTIVE))
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala:149]: Actually, can we have `def getIndexVersions(states: Seq[String]): Seq[Int]` instead of `getAvailableIndexVersions`? "Available" <=> "ACTIVE" doesn't seem to match nicely. (sorry for the wrong suggestion in the beginning). The caller `closestIndex` can decide which state it wants to retrieve.

File: src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala [hunk 7]
```diff
@@ -68,7 +68,9 @@ class DeltaLakeIntegrationTest extends QueryTest with HyperspaceSuite {
       dfFromSample.write.format("delta").save(dataPath)
 
       val deltaDf = spark.read.format("delta").load(dataPath)
-      hyperspace.createIndex(deltaDf, IndexConfig("deltaIndex", Seq("clicks"), Seq("Query")))
+      withSQLConf(IndexConstants.INDEX_LINEAGE_ENABLED -> "true") {
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala:71]: why is this required now?

**Review by imback82 — COMMENTED** [review]
LGTM (one minor comment), thanks @sezruby!

File: src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala (hunk 7 above)
Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala:71]: Since I added the condition above:
https://github.com/microsoft/hyperspace/pull/389/files/b938556191c3dcceab0ec784fa26defc5b02804b#diff-ccc638dd1aab8f054c6d13cdfafeea9c1634646e90dd999b491e4e43a18d9efbR188

It's possible to allow append-only hybrid scan if possible, but the condition in `closestIndex` will become more complicated. So just limit the case for now.

Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala:71]: Hmm, so before this PR, append-only hybrid scan was supported? I am curious which change in this PR triggered this new requirement.

Comment (sezruby) [reply src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala:71]: Might be supported. The index was not applied because of the conditions in getHybridScanCandidate if the version has deleted files (compared to the given relation) - before this PR. 

I found that I was writing a test for directory check change.

File: src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala [hunk 8]
```diff
@@ -194,9 +200,13 @@ class DeltaLakeRelation(spark: SparkSession, override val plan: LogicalRelation)
         .get
         .asInstanceOf[IndexLogEntry]
     }
-    // TODO: Currently assume all versions of index data exist.
-    //  Need to check and remove candidate indexes.
-    //  See https://github.com/microsoft/hyperspace/issues/387
+
+    // Get available index log versions.
+    val availableLogVersions =
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:205]: `activeLogVersions`? and I think you can remove the comment unless you want to explain "why" you are getting active logs.

File: src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala [hunk 9]
```diff
@@ -194,9 +200,13 @@ class DeltaLakeRelation(spark: SparkSession, override val plan: LogicalRelation)
         .get
         .asInstanceOf[IndexLogEntry]
     }
-    // TODO: Currently assume all versions of index data exist.
-    //  Need to check and remove candidate indexes.
-    //  See https://github.com/microsoft/hyperspace/issues/387
```
[the remark this answers is not in the document]
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:199]: I think we could close the issue now as index data existence is a different issue.

File: src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala [hunk 10]
```diff
@@ -184,8 +185,18 @@ class DeltaLakeRelation(spark: SparkSession, override val plan: LogicalRelation)
    * @return IndexLogEntry of the closest version among available index versions.
    */
   override def closestIndex(index: IndexLogEntry): IndexLogEntry = {
+
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:188]:
```suggestion
```

**Review by imback82 — APPROVED** [review]

**Review by imback82 — APPROVED** [review]

File: src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala:199]: sgtm
